docs/design/proposals/nfs.md

```diff
@@ -0,0 +1,65 @@
+# Dynamic provisioning of NFS volumes
+
+Ceph has [support for NFS-Ganesha to export directories][ceph_mgr_nfs] on
+CephFS. This can be used to export CephFS based volumes over NFS.
+
+## Node-Plugin for mounting NFS-exports
+
+The Kubernetes CSI community provides and maintains a [NFS CSI][nfs_csi]
+driver. This driver can be used as a Node-Plugin so that NFS CSI volumes can be
+mounted. When a CSI volume has the `server` and `share`
+[parameters][nfs_csi_params], the Node-Plugin will be able to mount the
+NFS-export.
+
+## Exporting CephFS based volumes over NFS
+
+Ceph-CSI already creates CephFS volumes, that can be mounted over the native
+CephFS protocol. A new provisioner in Ceph-CSI can create CephFS volumes, and
+include the required [NFS CSI parameters][nfs_csi_params] so that the [NFS CSI
+driver][nfs_csi] can mount the CephFS volume over NFS.
+
+The provisioner that handles the CSI requests for NFS volume can call the [Ceph
+Mgr commands to export/unexport][ceph_mgr_nfs] the CephFS volumes. The CephFS
+volumes would be internally managed by the NFS provisioner, and only be exposed
+as NFS CSI volumes towards the consumers.
+
+### `CreateVolume` CSI operation
+
+When the Ceph-CSI NFS provisioner is requested to create a NFS CSI volume, the
+following steps need to be taken:
+
+1. create a CephFS volume, use the CephFS `CreateVolume` call or other internal
+   API
+1. call the Ceph Mgr API to export the CephFS volume with NFS-Ganesha
+1. return the NFS CSI volume, with `server` and `share` parameters (other
+   parameters that are useful for CephFS volume management may be kept)
+
+The 2nd step requires a NFS-cluster name for the Ceph Mgr call(s). The name of
+the NFS-cluster as managed by Ceph should be provided in the parameters of the
+`CreateVolume` operation. For Kubernetes container orchistrators that means the
+parameters is set as an option in the `StorageClass`.
+
+The `server` parameter of the volume is an other option that is managed by the
+Ceph (or Rook) infrastructure. This parameter is also required to be provided
+in the `CreateVolume` parameters.
```

Comment:
What this `server` refers to ? if its mon IP there is a chance for the dynamic change . Or is it a static IP which always clients can connect to ? @nixpanic

Reply:
`server` is the parameter as mentioned for the NFS CSI NodePlugin, it should be the hostname (or IP-addresse) to the Ceph managed NFS-cluster.

Reply:
No, Ceph-CSI should not deploy the NFS-cluster, that is something Rook or the Ceph administrator should do. Ceph-CSI will just use Ceph Mgr to add new exported volumes to the existing Ceph managed NFS-Ganesha(s).

The `server` is therefore configured by Rook or an other operator (automated or human). It refers to the hostname that the NFS-cluster (or single NFS-Ganesha instance) is reachable by. A VIP, or some other load-balanced high-available mechanism is the same for the mounting clients, they do not care.

Reply:
well, this server IP or server address is static and can not be changed, so the concern . If upper layers manage it properly, it should work though.